GeeksCoreLibrary/Modules/Payments/Services/BuckarooService.cs

```diff
@@ -252,6 +285,66 @@ public async Task<StatusUpdateResult> ProcessStatusUpdateAsync(OrderProcessSetti
             };
         }
 
+        /// <summary>
+        /// Handles the status update using form values.
+        /// </summary>
+        /// <param name="buckarooSettings">The settings for Buckaroo.</param>
+        /// <param name="invoiceNumber">The payment's invoice number.</param>
+        /// <returns>A <see cref="StatusUpdateResult"/> object.</returns>
+        private async Task<StatusUpdateResult> HandleFormStatusUpdate(BuckarooSettingsModel buckarooSettings, string invoiceNumber)
+        {
+            if (httpContextAccessor.HttpContext == null)
+            {
+                return new StatusUpdateResult
+                {
+                    Status = "No HTTP context available; unable to process status update.",
+                    Successful = false
+                };
+            }
+
+            var statusCode = httpContextAccessor.HttpContext.Request.Form["brq_statuscode"].ToString();
+
+            // Get all form values that begin with "brq_", "add_" or "cust_", except "brq_signature".
+            var formValues = httpContextAccessor.HttpContext.Request.Form.Where(kvp => !kvp.Key.Equals("brq_signature") && (kvp.Key.StartsWith("brq_") || kvp.Key.StartsWith("add_") || kvp.Key.StartsWith("cust_"))).ToDictionary(kvp => kvp.Key, kvp => kvp.Value.ToString());
+            var buckarooSignature = httpContextAccessor.HttpContext.Request.Form["brq_signature"].ToString();
+
+            // Sort the formValues dictionary alphabetically by key.
+            formValues = formValues.OrderBy(kvp => kvp.Key).ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+            var signatureBuilder = new StringBuilder();
+            foreach (var formValue in formValues)
+            {
+                signatureBuilder.Append($"{formValue.Key}={formValue.Value}");
+            }
+
+            if (!String.IsNullOrWhiteSpace(buckarooSettings.SecretKey))
+            {
+                signatureBuilder.Append(buckarooSettings.SecretKey);
+            }
+
+            // Hash the signature builder with SHA1.
+            var hash = SHA1.HashData(Encoding.UTF8.GetBytes(signatureBuilder.ToString()));
+            var signatureHash = BitConverter.ToString(hash).Replace("-", "").ToLower();
+
+            // Compare hashes.
+            if (String.Equals(buckarooSignature, signatureHash, StringComparison.OrdinalIgnoreCase))
+            {
+                return new StatusUpdateResult
+                {
+                    Status = httpContextAccessor.HttpContext.Request.Form["brq_statusmessage"].ToString(),
+                    Successful = statusCode.InList("190", "790")
+                };
+            }
+
+            await LogIncomingPaymentActionAsync(PaymentServiceProviders.Buckaroo, invoiceNumber, Int32.TryParse(statusCode, out var status) ? status : 0, signatureBuilder.ToString());
```

Comment:
Deze logging gaat volgens mij niet helemaal goed. Deze wordt nu alleen gedaan wanneer de signature ongeldig is. Je wilt inkomende verzoeken altijd loggen, lijkt me. Daarnaast log je de inhoud van signatureBuilder, maar dat is een waarde die jezelf in code opbouwt. Je wilt in de logs exact zien wat Buckaroo heeft gestuurd, niet wat wij er van hebben gemaakt.

Daarnaast lijkt me dat deze logging beter in `ProcessStatusUpdateAsync` kan komen in een try/finally, zodat we altijd loggen wat er gebeurt, ipv alleen in bepaalde situaties.

Reply:
Ik heb de logging uitgebreid zodat er altijd wordt gelogd middels een `try-catch` (waar de logging in het `finally` blok zit).